### PR TITLE
`generic_float` for Float8E8M0

### PR DESCRIPTION
**Pull request by CharlieL7**
## Motivation

* Introduce Float8E8M0 type within MIGraphX for better MXFP4 optimizations and to use hipblaslt mxfp4 kernels.

## Changelog Category

- - [ ] Added: New functionality.
- - [ ] Changed: Changes to existing functionality.
- - [ ] Removed: Functionality or support that has been removed. (Compared to a previous release)
- - [ ] Optimized: Component performance that has been optimized or improved.
- - [ ] Resolved Issues: Known issues from a previous version that have been resolved.
- - [ ] Not Applicable: This PR is not to be included in the changelog.

### CODE REVIEW

File: src/include/migraphx/generic_float.hpp [hunk 1]
```diff
@@ -373,7 +531,7 @@ template <unsigned int M, unsigned int E, unsigned int F>
 class numeric_limits<migraphx::generic_float<M, E, F>>
 {
     public:
-    static constexpr bool has_infinity = true;
+    static constexpr bool has_infinity = not(M == 0 and E == 0);
```
Comment (TedThemistokleous) [inline src/include/migraphx/generic_float.hpp:534]: Tidy is being nitpicky here but It saves you the extra not if you make this  `(M == 0 or E == 0)`

File: src/netron_output.cpp [hunk 2]
```diff
@@ -57,7 +57,8 @@ int get_onnx_type(shape::type_t s_type)
     case shape::fp8e5m2_type: return 19;
     case shape::fp8e5m2fnuz_type: return 20;
     case shape::tuple_type: return 0;
-    case shape::fp4x2_type: return 21; // TODO update this when the type is added
+    case shape::fp4x2_type: return 23;
```
Comment (TedThemistokleous) [inline src/netron_output.cpp:60]: Not sure why this is no 23 from 21? Shouldn't these be the same and not change or is this more related to some other representation of things?

**Comment by TedThemistokleous** [comment]
So far so good - Tidy seems a bit heavy handed here and CI is giving a nonsensical error for some reason across builds

```
[2025-10-23T22:49:15.868Z] /var/jenkins/workspace/AMDMIGraphX_PR-4403/src/targets/gpu/hip_gemm_impl.cpp:77:5: error: label at end of compound statement is a C++23 extension [-Werror,-Wc++23-extensions]

[2025-10-23T22:49:15.868Z]    77 |     }

[2025-10-23T22:49:15.868Z]       |     ^

[2025-10-23T22:49:15.868Z] 1 error generated.

[2025-10-23T22:49:15.868Z] make[2]: *** [src/targets/gpu/CMakeFiles/migraphx_gpu.dir/build.make:370: src/targets/gpu/CMakeFiles/migraphx_gpu.dir/hip_gemm_impl.cpp.o] Error 1

[2025-10-23T22:49:15.868Z] make[2]: *** Waiting for unfinished jobs....

[2025-10-23T22:49:15.868Z] /var/jenkins/workspace/AMDMIGraphX_PR-4403/src/targets/gpu/gemm_impl.cpp:79:5: error: label at end of compound statement is a C++23 extension [-Werror,-Wc++23-extensions]

[2025-10-23T22:49:15.868Z]    79 |     }

[2025-10-23T22:49:15.868Z]       |     ^

[2025-10-23T22:49:15.868Z] 1 error generated.

[2025-10-23T22:49:15.868Z] make[2]: *** [src/targets/gpu/CMakeFiles/migraphx_gpu.dir/build.make:328: src/targets/gpu/CMakeFiles/migraphx_gpu.dir/gemm_impl.cpp.o] Error 1

[2025-10-23T22:49:30.250Z] make[2]: Leaving directory '/var/jenkins/workspace/AMDMIGraphX_PR-4403/build'

[2025-10-23T22:49:30.251Z] make[1]: *** [CMakeFiles/Makefile2:21332: src/targets/gpu/CMakeFiles/migraphx_gpu.dir/all] Error 2

[2025-10-23T22:49:30.251Z] make[1]: Leaving directory '/var/jenkins/workspace/AMDMIGraphX_PR-4403/build'

```

Let me know when you want me to look at this again and its ready for full review unless this needs to be taken out of draft. Not really issues with this right now if that's the case

File: src/include/migraphx/generic_float.hpp (hunk 1 above)
Comment (CharlieL7) [reply src/include/migraphx/generic_float.hpp:534]: That's actually a bug. Should probably just be `M != 0`.

File: src/netron_output.cpp (hunk 2 above)
Comment (CharlieL7) [reply src/netron_output.cpp:60]: There are: 
```
20: onnx.TensorProto.FLOAT8E5M2FNUZ
21: onnx.TensorProto.UINT4
22: onnx.TensorProto.INT4
23: onnx.TensorProto.FLOAT4E2M1
24: onnx.TensorProto.FLOAT8E8M0
```
It just so happens that the int4 types are not used.

Comment (TedThemistokleous) [reply src/netron_output.cpp:60]: Ah understood

File: src/include/migraphx/generic_float.hpp (hunk 1 above)
Comment (pfultz2) [reply src/include/migraphx/generic_float.hpp:534]: Tidy is not suggesting `(M == 0 or E == 0)`, it is suggesting `M != 0 or E != 0`.

File: src/include/migraphx/generic_float.hpp [hunk 3]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
     }
 };
 
+template <unsigned int Flags>
+struct __attribute__((packed, may_alias)) generic_float<0, 8, Flags>
+    : generic_float_operators<generic_float<0, 8, Flags>>
+{
+    uint8_t exponent;
+
+    static constexpr int exponent_bias() { return all_ones<7>(); }
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:388]: Nit: The `bias` is just an integer, per the standard. it should ideally be just referred to by its value: `127`, or something like that (a named constant). But defining it all_ones<7> is confusing.

File: src/include/migraphx/generic_float.hpp [hunk 4]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
     }
 };
 
+template <unsigned int Flags>
+struct __attribute__((packed, may_alias)) generic_float<0, 8, Flags>
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:383]: Is `Flags` used somewhere in E8M0 case here? I must have missed.

File: src/include/migraphx/generic_float.hpp [hunk 5]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
...
+    : generic_float_operators<generic_float<0, 8, Flags>>
+{
+    uint8_t exponent;
+
+    static constexpr int exponent_bias() { return all_ones<7>(); }
+
+    explicit constexpr generic_float(float f = 1.0) noexcept { from_float(get_parts(f)); }
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:390]: Would you be defining a `generic_float` for other formats with a different default value?

File: src/include/migraphx/generic_float.hpp [hunk 6]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
...
+
+    constexpr generic_float operator+() const noexcept { return *this; }
+
+    constexpr float to_float() const noexcept
+    {
+        float32_parts f{};
+        f.sign     = 0;
+        if(exponent == 0)
+        {
+            // 2^(-127) is a fp32 denormal number
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:409]: `2 ** -127` is not denormal in FP32, AFAIK.

File: src/include/migraphx/generic_float.hpp [hunk 7]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
...
+        ;
+    }
+
+    constexpr operator float() const noexcept { return this->to_float(); }
+
+    // doesn't have infinity, returning 2**0
+    static constexpr generic_float infinity()
+    {
+        generic_float x{};
+        x.exponent = all_ones<8>() >> 1u;
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:455]: ~~This I don't get. In `fp32`, `infinity` has its exponent defined as `0xff`. why is the above line shifting right?~~  Looks like you are converting infinity to the max() here. Maybe just use max() in that case?

File: src/include/migraphx/generic_float.hpp (hunk 6 above)
Comment (CharlieL7) [reply src/include/migraphx/generic_float.hpp:409]: FLT_MIN is 1.17549e-38 which 2**(-126). 2**(-127) is represented in fp32 as exponent = 0 and mantissa = (1 << 22).

File: src/include/migraphx/generic_float.hpp (hunk 7 above)
Comment (CharlieL7) [reply src/include/migraphx/generic_float.hpp:455]: There is no infinity for E8M0 so I tried to follow the convention integral types have for this https://en.cppreference.com/w/cpp/types/numeric_limits/infinity.html. Unfortunately, there is also no zero in E8M0 so this decision is arbitrary.

File: src/include/migraphx/generic_float.hpp [hunk 8]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
...
+        return x;
+    }
+
+    // only one NaN value
+    static constexpr generic_float qnan() { return snan(); }
+
+    // min value = 2**(-127)
+    static constexpr generic_float min()
+    {
+        generic_float x{};
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:473]: Why not define `generic_float` natively -- rather than go through a default `fp32` parameter in its constructor, subsequently its exponent is set!

File: src/include/migraphx/generic_float.hpp [hunk 9]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
...
+    // max value = 2**(127)
+    static constexpr generic_float max()
+    {
+        generic_float x{};
+        x.exponent = all_ones<8>() - 1;
+        return x;
+    }
+
+    // next number from 2**0 is 2**1 so epsilon is 2**0
+    static constexpr generic_float epsilon()
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:492]: If I were to nit-pick, there is no `epsilon` or a concept like that for the weight specific E8M0 format.

File: src/include/migraphx/generic_float.hpp [hunk 10]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
...
+    {
+        generic_float x{};
+        x.exponent = all_ones<8>() >> 1u;
+        return x;
+    }
+
+    friend constexpr bool operator==(const generic_float& x, const generic_float& y)
+    {
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:500]: Why not just say `x == y`. Is that not compliable?

File: src/include/migraphx/generic_float.hpp [hunk 11]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
...
+    }
+
+    friend constexpr bool operator!=(const generic_float& x, const generic_float& y)
+    {
+        return not(x == y);
+    }
+
+    constexpr generic_float& operator++() noexcept
+    {
+        ++exponent;
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:512]: You perhaps should check for a number to not become a `NaN`. It should roll over.

File: src/include/migraphx/generic_float.hpp (hunk 6 above)
Comment (lakhinderwalia) [reply src/include/migraphx/generic_float.hpp:409]: You are correct. Thanks.

File: src/include/migraphx/generic_float.hpp (hunk 7 above)
Comment (lakhinderwalia) [reply src/include/migraphx/generic_float.hpp:455]: I agree there is no `zero` involved here :-)
Should an exception be generated? Because it is the wrong math otherwise for E8M0. 
Ideally the test could verify that an exception is being thrown.
These are hairy corner cases.

File: src/include/migraphx/generic_float.hpp [hunk 12]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
...
+    explicit constexpr generic_float(float f = 1.0) noexcept { from_float(get_parts(f)); }
+
+    constexpr generic_float& operator=(float f) noexcept
+    {
+        from_float(get_parts(f));
+        return *this;
+    }
+
+    // No sign for this type
+    constexpr generic_float operator-() const noexcept { return snan(); }
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:399]: If there is no unary `-` op available, then logically speaking a binary `-` or `-=` op is also the same category. One could do a binary `x - y`, which is say `2 - 2`, and get some strange results!

Comment (lakhinderwalia) [reply src/include/migraphx/generic_float.hpp:399]: Also, fundamentally, `a - b` == `a` `+` `(` `-` `b` `)`

File: src/include/migraphx/generic_float.hpp [hunk 13]
```diff
@@ -363,6 +379,148 @@ struct __attribute__((packed, may_alias)) generic_float
...
+    : generic_float_operators<generic_float<0, 8, Flags>>
+{
+    uint8_t exponent;
+
+    static constexpr int exponent_bias() { return all_ones<7>(); }
+
+    explicit constexpr generic_float(float f = 1.0) noexcept { from_float(get_parts(f)); }
+
+    constexpr generic_float& operator=(float f) noexcept
+    {
```
Comment (lakhinderwalia) [inline src/include/migraphx/generic_float.hpp:393]: Should this assign from a negative float value?

File: src/include/migraphx/generic_float.hpp (hunk 12 above)
Comment (lakhinderwalia) [reply src/include/migraphx/generic_float.hpp:399]: If you can allow an assignment from a negative float, then `operator-()` should just be a `nop`, but not a `snan()`.  But that totally messes up `a - b`. 

Therefore, no assignment from a negative value should be allowed. 

An interesting problem :-)

File: src/include/migraphx/generic_float.hpp (hunk 13 above)
Comment (CharlieL7) [reply src/include/migraphx/generic_float.hpp:393]: It can, I set it up to ignore the sign bit.

File: src/include/migraphx/generic_float.hpp (hunk 12 above)
Comment (CharlieL7) [reply src/include/migraphx/generic_float.hpp:399]: There should probably only be the equality binary operation.

File: src/include/migraphx/generic_float.hpp (hunk 13 above)
Comment (lakhinderwalia) [reply src/include/migraphx/generic_float.hpp:393]: I meant logically speaking, the answer to this question should be: "negative" :-)

File: src/include/migraphx/fp8e8m0.hpp [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+#ifndef MIGRAPHX_GUARD_RTGLIB_FP8E8M0FN_HPP
+#define MIGRAPHX_GUARD_RTGLIB_FP8E8M0FN_HPP
+
+#include <migraphx/generic_float.hpp>
+#include <migraphx/config.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+using fp8e8m0 = migraphx::generic_float<0, 8, 0>;
```
Comment (pfultz2) [inline src/include/migraphx/fp8e8m0.hpp:34]: A template parameter needs to be added for the sign. So it should become:

```cpp
template <unsigned int MantissaSize, unsigned int ExponentSize, bool Signed = true, unsigned int Flags = 0>
struct __attribute__((packed, may_alias)) generic_float;
```

And the typedef should then be `using fp8e8m0 = migraphx::generic_float<0, 8, false>`.

**Comment by kahmed10** [comment]
lower priority since there are no models that support this datatype currently
